proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,554 @@
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+.. sectnum::
+.. contents::
+
+Most languages have native support for writing strings with newlines. Rust supports it out of the box with the normal ``"`` syntax, Python and Scala have ``"""`` syntax, and Javascript/Typescript has backticks. This proposal recommends adding ``"""`` as a delimiter which starts and ends a multiline string that automatically strips the common whitespace prefix from lines.
+
+This proposal is heavily influenced by Java's text blocks in `JEP 378 <https://openjdk.org/jeps/378>`_.
+
+Motivation
+----------
+
+Currently you could do this with ``unlines`` (more commonly) or line continuations (`"gaps" <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-200002.6>`_):
+
+::
+
+  s1 = unlines
+    [ "line 1"
+    , "line 2"
+    , "line 3"
+    ]
+
+  s1_2 = "line 1\n\
+         \line 2\n\
+         \line 3\n"
+
+Using ``unlines`` is decent, but it's a bit verbose. The verbosity becomes more apparent when building multiline strings with double quotes, e.g. rendering Haskell code or SQL queries.
+
+::
+
+  s2 = unlines
+    [ "x :: String"
+    , "x = a ++ b"
+    , "  where"
+    , "    a = \"Hello \""
+    , "    b = \"world!\""
+    ]
+
+  s3 = unlines
+    [ "SELECT \"id\", \"name\""
+    , "FROM \"user\""
+    , "WHERE \"name\" ILIKE 'Alice%'"
+    ]
+
+Compare with multiline strings:
+
+::
+
+  s1' =
+    """
+    line 1
+    line 2
+    line 3
+    """
+
+  s2' =
+    """
+    x :: String
+    x = a ++ b
+      where
+        a = "Hello "
+        b = "world!"
+    """
+
+  s3' =
+    """
+    SELECT "id", "name"
+    FROM "user"
+    WHERE "name" ILIKE 'Alice%'
+    """
+
+Third party libraries also provide this functionality with quasiquoters, e.g. ``heredoc`` or libraries that also do interpolation like ``string-interpolate``. But Template Haskell is not great:
+
+* It makes compilation difficult in certain environments (e.g. GHCJS)
+* It can slow down compilation
+* Some people avoid it as much as possible out of principle
+* It's a rather heavyweight tool for a seemingly lightweight syntactic feature
+
+Proposed Change Specification
+-----------------------------
+
+#. Lex ``"""`` as an `additional string delimiter <https://gitlab.haskell.org/ghc/ghc/-/blob/8c0ea25fb4a27d4729aabf73f4c00b912bb0c58d/compiler/GHC/Parser/Lexer.x#L577>`_ when the ``MultilineStrings`` language extension is enabled
+
+#. Post-process the string in the following steps:
+
+   #. If the string starts with a newline character, ignore it
+
+   #. Collapse line continuations
+
+   #. Remove common whitespace prefix in every line
+
+      * Ignore any characters preceding the first newline
+      * Blank lines and lines with only whitespace should not be included in this calculation
+
+   #. Remove exactly one newline from the beginning of the string (if one exists)
```

Comment:
IIUC, the first and last steps here are trying to do the same thing, and only the last one should be kept. (Needs to come after the common prefix removal.)

Reply:
Yes, thanks!